### PR TITLE
DALI for aarch64-linux platform

### PR DESCRIPTION
**Pull request by klecki**
Based on #522 - squashed & rebased on current master (522 is the first commit here).

- [x] Run the tests on aarch64
- [x] Clean up the FindCuda parts from logs
- [x] Adjust the CMake commands in docker to recent changes
- [x] Sign-off the commits...
~~Propagate fixes for usage of FindX from aarch64-linux to qnx.~~
~~Build & test QNX after changes~~

QNX is moved to separate PR #890

### CODE REVIEW

**Comment by klecki** [comment]
!build

File: cmake/Dependencies.aarch64-linux.cmake [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+message(STATUS "CMAKE_CROSSCOMPILING ${CMAKE_CROSSCOMPILING}")
+
+
+set(CUDA_TOOLKIT_ROOT_DIR ${CUDA_HOST})
+set(CUDA_TOOLKIT_TARGET_DIR ${CUDA_TARGET})
+
+find_package(CUDA 10.0 REQUIRED)
+message(STATUS "CUDA_VARIABLES:")
+
+# message(STATUS "CUDA_VERSION_MAJOR: ${CUDA_VERSION_MAJOR}")
```
Comment (andi4191) [inline cmake/Dependencies.aarch64-linux.cmake]: Clean up for commented part?

File: cmake/Dependencies.aarch64-linux.cmake [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+message(STATUS "CUDA_nppc_static_LIBRARY: ${CUDA_nppc_static_LIBRARY}")
+message(STATUS "CUDA_nppicom_static_LIBRARY: ${CUDA_nppicom_static_LIBRARY}")
+message(STATUS "CUDA_nppicc_static_LIBRARY: ${CUDA_nppicc_static_LIBRARY}")
+message(STATUS "CUDA_nppig_static_LIBRARY: ${CUDA_nppig_static_LIBRARY}")
+message(STATUS "CUDA_culibos_LIBRARY: ${CUDA_culibos_LIBRARY}")
+
+include_directories(${CUDA_TOOLKIT_TARGET_DIR}/include)
+include_directories(${CUDA_TOOLKIT_ROOT_DIR}/include)
+
+# set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -L${CUDA_TARGET_LIBRARIES_DIR} -L${CUDA_TARGET_LIBRARIES_DIR}/stubs -lcudart -lnppc_static -lnppicom_static -lnppicc_static -lnppig_static -lculibos")
```
Comment (andi4191) [inline cmake/Dependencies.aarch64-linux.cmake]: clean up?

File: cmake/Dependencies.cmake [hunk 3]
```diff
@@ -48,8 +67,8 @@ if (BUILD_NVJPEG)
     add_definitions(-DNVJPEG_DECOUPLED_API)
   endif()
 else()
-  # Note: Support for disabling nvJPEG is unofficial
-  message(STATUS "Building WITHOUT nvJPEG")
+    # Note: Support for disabling nvJPEG is unofficial
```
Comment (andi4191) [inline cmake/Dependencies.cmake]: indentation out of place

File: dali/kernels/alloc.cc [hunk 4]
```diff
@@ -91,7 +92,7 @@ void Deallocate(AllocType type, void *mem, int device) noexcept {
   VALUE_SWITCH(type, type_label,
     (AllocType::Host, AllocType::Pinned, AllocType::GPU, AllocType::Unified),
     (return Allocator<type_label>::Deallocate(mem, device)),
-    (assert(!"Invalid allocation type requested");)
+    (assert(!"Invalid Deallocation type requested");)
```
Comment (mzient) [inline dali/kernels/alloc.cc]: Why? There's no such thing as "deallocation type". However, it's not _requested_, so maybe:
```suggestion
    (assert(!"Invalid allocation type specified");)
```

Comment (klecki) [reply dali/kernels/alloc.cc]: Missed it, will revert to how it was before.

File: cmake/Dependencies.aarch64-linux.cmake (hunk 2 above)
Comment (klecki) [reply cmake/Dependencies.aarch64-linux.cmake]: Yes, done.

File: cmake/Dependencies.cmake (hunk 3 above)
Comment (klecki) [reply cmake/Dependencies.cmake]: Done

File: cmake/Dependencies.aarch64-linux.cmake (hunk 1 above)
Comment (klecki) [reply cmake/Dependencies.aarch64-linux.cmake]: Done.

**Comment by klecki** [comment]
!build

File: dali/kernels/alloc.cc (hunk 4 above)
Comment (andi4191) [reply dali/kernels/alloc.cc]: I misunderstood it with the function name instead of parameter name.

**Comment by klecki** [comment]
!build

File: cmake/Dependencies.cmake [hunk 5]
```diff
@@ -153,48 +100,13 @@ else()
   add_definitions(-DDALI_BUILD_PROTO3=1)
   set(BUILD_PROTO3 ON CACHE STRING "Build proto3")
 endif()
-include_directories(SYSTEM ${PROTOBUF_INCLUDE_DIRS})
-list(APPEND DALI_LIBS ${PROTOBUF_LIBRARY})
+
+include_directories(SYSTEM ${Protobuf_INCLUDE_DIRS})
+list(APPEND DALI_LIBS ${Protobuf_LIBRARY})
 # hide things from the protobuf, all we export is only is API generated from our proto files
 list(APPEND DALI_EXCLUDES libprotobuf.a)
```
Comment (klecki) [inline cmake/Dependencies.cmake]: TODO: Exclude it for aarch64 as well.

File: dali/pipeline/pipeline.cc [hunk 6]
```diff
@@ -66,7 +67,14 @@ namespace dali {
                      int default_cuda_stream_priority)
       : built_(false), separated_execution_(false) {
     dali_proto::PipelineDef def;
-    def.ParseFromString(serialized_pipe);
+    //  Reading Protobuf file has a limitation of 64 MB
+    //  Following instructions will increase the
+    //  limits to support the given pipeline
+    google::protobuf::io::CodedInputStream *coded_input =
+        new google::protobuf::io::CodedInputStream(
```
Comment (jantonguirao) [inline dali/pipeline/pipeline.cc]: aren't you deleting this?

File: Dockerfile.build.aarch64-linux [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+    unzip \
+    git \
+    rsync \
+    libjpeg-dev \
+    dh-autoreconf \
+    gcc-aarch64-linux-gnu \
+    g++-aarch64-linux-gnu \
+    && rm -rf /var/lib/apt/lists/*
+
+RUN wget https://developer.download.nvidia.com/devzone/devcenter/mobile/jetpack_l4t/4.1.1/xddsn.im/JetPackL4T_4.1.1_b57/16.04/cuda-repo-ubuntu1604-10-0-local-10.0.117-410.38_1.0-1_amd64.deb
```
Comment (JanuszL) [inline Dockerfile.build.aarch64-linux]: I would make L16-L32 a one Docker step as we don't need intermediate layers.

File: README.rst [hunk 8]
```diff
@@ -362,7 +362,42 @@ Install Python bindings
...
+-----------------------------------------
+Build the aarch64 Linux Build Container 
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+.. code-block:: bash
+    docker build -t dali_builder:aarch64-linux -f Dockerfile.build.aarch64-linux .
+
+Compile
+^^^^^^^
+From the root of the DALI source tree
+.. code-block:: bash
```
Comment (JanuszL) [inline README.rst]: You need a newline before the `.. code-block:: bash`

File: README.rst [hunk 9]
```diff
@@ -362,7 +362,42 @@ Install Python bindings
...
+Setup
+^^^^^
+After aquiring the QNX Toolchain, place it in a directory called ``qnx`` in the root of the DALI tree.
+Then using the SDK Manager for NVIDIA DRIVE, select **QNX** as the *Target Operating System* and select **DRIVE OS 5.1.0.0 SDK**
+In STEP 02 under **Download & Install Options**, select *Download Now. Install Later*. and agree to the Terms and Conditions. 
+Once downloaded move the **cuda-repo-cross-qnx** debian package into the ``qnx`` directory you created in the DALI tree. 
+
+Build the aarch64 Build Container 
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+.. code-block:: bash
```
Comment (JanuszL) [inline README.rst]: You need a newline before the `.. code-block:: bash`

File: README.rst [hunk 10]
```diff
@@ -362,7 +362,42 @@ Install Python bindings
...
+    docker build -t dali_builder:aarch64-qnx -f Dockerfile.build.aarch64-qnx .
+
+Compile
+^^^^^^^
+From the root of the DALI source tree
+.. code-block:: bash
```
Comment (JanuszL) [inline README.rst]: You need a newline before the `.. code-block:: bash`

File: dali/util/local_file.cc [hunk 11]
```diff
@@ -36,18 +38,18 @@
 static int _sysctl(struct __sysctl_args *args);
 
 static int get_max_vm_cnt() {
-  struct __sysctl_args args = {0, };
-  int vm_cnt;
+  int vm_cnt = 1;
+#if !defined(__AARCH64_GNU__)
```
Comment (JanuszL) [inline dali/util/local_file.cc]: Doesn't AARCH64 have `/proc/sys/vm/max_map_count`? Maybe we can fallback to L54 immediately for it?

Comment (klecki) [reply dali/util/local_file.cc]: It does, I can change it.

File: dali/pipeline/pipeline.cc (hunk 6 above)
Comment (andi4191) [reply dali/pipeline/pipeline.cc]: @jantonguirao: coded_input is being pased into Pipeline def here. Are you referring to deallocation of coded_input?

File: README.rst (hunk 8 above)
Comment (klecki) [reply README.rst]: Done

File: dali/util/local_file.cc (hunk 11 above)
Comment (klecki) [reply dali/util/local_file.cc]: Done

File: dali/pipeline/pipeline.cc (hunk 6 above)
Comment (klecki) [reply dali/pipeline/pipeline.cc]: Done

**Comment by klecki** [comment]
!build

**Comment by klecki** [comment]
!build

**Comment by klecki** [comment]
!build

File: cmake/Dependencies.cmake (hunk 5 above)
Comment (klecki) [reply cmake/Dependencies.cmake]: Done

File: Dockerfile.build.aarch64-linux [hunk 12]
```diff
@@ -0,0 +1,151 @@
+FROM nvidia/cuda:10.0-devel-ubuntu16.04
```
Comment (JanuszL) [inline Dockerfile.build.aarch64-linux]: Does it need to be nvidia/cuda:10.0-devel-ubuntu16.04 as we are installing Jetpack on our own anyway?

File: README.rst [hunk 13]
```diff
@@ -362,7 +362,50 @@ Install Python bindings
 
 .. installation-end-marker-do-not-remove
 
-----
+Cross-compiling DALI C++ API for aarch64 Linux (Docker)
+-----------------------------------------
```
Comment (JanuszL) [inline README.rst]: Underline should match string length.

File: README.rst [hunk 14]
```diff
@@ -362,7 +362,50 @@ Install Python bindings
...
+From the root of the DALI source tree
+
+.. code-block:: bash
+
+    docker run -v $(pwd):/dali dali_builder:aarch64-linux
+
+The relevant artifacts will be in ``build/install`` and ``build/dali/python/nvidia/dali``
+
+Cross-compiling DALI C++ API for aarch64 QNX (Docker)
+-----------------------------------------
```
Comment (JanuszL) [inline README.rst]: Too short undeline.

File: Dockerfile.build.aarch64-linux (hunk 12 above)
Comment (klecki) [reply Dockerfile.build.aarch64-linux]: I'm not sure if that package that we use for install provides the whole SDK or only the cross libraries.

File: README.rst (hunk 13 above)
Comment (klecki) [reply README.rst]: Done

File: README.rst (hunk 14 above)
Comment (klecki) [reply README.rst]: Done

**Pull request merged**